### ISSUE TITLE
Discussion: Making the gateway obsolete

### ISSUE THREAD

**Issue by JackD111** [Voyz/ibeam#50]
Since there is no other forum to discuss this I figured an "enhancement" issue might be the best idea.

So recently I was thinking why we even have to use the IB gateway client in the first place. Why not just run a headless browser to authenticate and use the cookies after a successfull auth to run against the IB client portal api?

So i just tested this really quick and as expected it actually works that easy. Open your reginal IBKR site, authenticate and after the out you grab the `USERID` cookie that will get responed from the server from the `/sso/Dispatcher`. That cookie seems to represent some kind of session since that seems to be the only thing you need to fire requests to the `portal.proxy`. With the cookie you essentially just do requests against the portal proxy which for my region is under `https://www.interactivebrokers.co.uk/portal.proxy/v1/portal`. The IB client portal runs all its requets against that and a quick check shows the same responses as the IB gateway would give you.

Going this route would make the IB gateway obsolete and we could make direct requests to the `portal.proxy` endpoint.
Any toughts?

**Comment by 10ne1** [comment]
This is an awesome idea if it works reliably. Thanks for suggesting it, I'll definitely start testing on my own end (if only I had more time...).

If not in ibeam, maybe we could create a separate project.

I'm not exactly sure which cookies are required, but just to be safe I downloaded all session cookies from the IBKR website into a `cookies.txt`, then ran for eg the following with the proper `<user-id>` replaced:

```
curl -b cookies.txt https://www.interactivebrokers.co.uk/portal.proxy/v1/portal/portfolio/<user-id>/summary
```

Some observations:

1. It is much faster and more deterministic than routing through the gateway (Java runtime is garbege, so it's expected).

2. I don't think the "region" matters for anything as I login from .hu then get redirected to .co.uk anyway.

3. Sneaked a look into the gateway code. It is as bad as expected. Just more reason to get rid of it.

4. Making the gateway obsolete will also make significant parts of the docker container obsolete so it will at least decrease in size (good because VPS space is expensive). Not sure if it still makes sense to ship ibeam within a container without the gateway, but it's still a significant plus IMO.

Now that I know this works I'm tempted to just plug this authentication into my Rust + serde + reqwest codebase and do away with ibeam entirely, see how it goes. :joy:

**Comment by JackD111** [comment]
Yeah lots of reasons to ditch the gateway. Only the `USERID` cookie is required. Seems to be some kind of session cookie.

Something like this would require a new project but I wanted to bounce of the general idea of this new approach on people

**Comment by 10ne1** [comment]
Good news: All endpoints (the ones which my app uses anyway :smile: ) seem to work with just two cookies: `USERID` and `XYZAB`.

`USERID` appears to be a constant, per-user, but different than the account UID used to identify trader accounts in the IBKR database.

`XYZAB` looks like a SHA-1 checksum and changes at every login. It is identical to `XYZAB_AM.LOGIN`. This seems to be the actual session-identification cookie.

Anyway, ibeam could be converted to this new interface and continue handling the browser (re-)auth, tickle, maybe it would even help with those annoying "Gateway running but not authenticated" errors, not sure if we need to start a new project...

If it helps, here's my session setup logic in Rust, I'm still hardcoding the cookies because I haven't yet added chromedriver communication.

```Rust
pub fn get_ibkr_client() -> reqwest::Client {
    let userid = "...";
    let xyzab = "...";
    let domain = ".interactivebrokers.co.uk";

    let cookie1 = format!("USERID={}; Domain={}", userid, domain);
    let cookie2 = format!("XYZAB={}; Domain={}", xyzab, domain);

    let url = format!("https://www{}", domain)
        .parse::<Url>()
        .expect("Failed to parse IBKR URL");

    let jar = Jar::default();
    jar.add_cookie_str(cookie1.as_str(), &url);
    jar.add_cookie_str(cookie2.as_str(), &url);

    reqwest::Client::builder()
        .cookie_store(true)
        .cookie_provider(Arc::new(jar))
        .build()
        .expect("Failed to build IBKR reqwest client")
}
```

Then where needed I do calls like

```Rust
ibkr_client
    .post("https://www.interactivebrokers.co.uk/portal.proxy/v1/portal/trsrv/secdef")
    .header("Content-Type", "application/json")
    .body(all_conid_json)
    .send()
```

One more thing, I think that having an intermediary web service doing validation is unnecessary, since we called these end-points directly via the gateway as well. The latency is quite decent now, why introduce a new intermediary when we just got rid of the Java gateway? :wink:

**Comment by Voyz** [comment]
Hey @JackD111 and @10ne1 thanks for making this suggestion and already sharing progress in this area. 

I will note that there has been similar attempts made by other users in the past, and I remember that IBKR has warned them such actions go against their TOC and could result in unwanted consequences from their side. I would encourage you to communicate with IBKR before proceeding with this approach and sharing your results publicly. 

If IBKR would be okay to go ahead then you're making some great contribution here, I'd just double check it's okay to do so 👍

**Comment by 10ne1** [comment]
Thanks for the info @Voyz . I've stopped my research until we can get some confirmation that it's ok to call the IBKR API without the gateway.

**Comment by JackD111** [comment]
I've also started working on an API wrapper but for .NET that can already do the headless auth and run some basic requests that I've thrown in together in the weekend.

I've opened a ticket with support and asked if it was allowed or against their ToS or anything. Will get back to you guys about this once I get an answer.

Hopefully they give a postive answer back since I'd really rather like to work indipendant of the IB gateway. I  mean in the end they wouldn't be able to distinigsh API requests from an actual browser request vs. code requests when sending the proper cookies/headers unless you run your trading bot in a data center like me and than compare the IP addresses to known data centers but I heavily doubt they'd go so far.

**Comment by 10ne1** [comment]
Thanks @JackD111 I also got auth working yesterday with the Rust `fantoccini` crate, but that's about it. Until we hear back from IBKR I'm stopping at this point. I also need to focus on actual work, so unfortunately I don't have much time to spend on this.

**Comment by stopdesign** [comment]
@JackD111
> issues with the gateway client today

Yes, it was offline since ~02:15 UTC. I've got a lot of different errors, not only http-500.
I thought the gateway was finally dead.

The regular (not web api) IB Gateway is much more reliable but also have issues. So I second the idea of using the `portal.proxy` directly.

**Comment by 10ne1** [comment]
The web API, website and mobile app regularly have issues, some endpoints might not work at all, return random errors, even bogus conid data. The mobile app displays this notification:

```
IMPORTANT: Due to scheduled maintenance clients may be unable to connect through our web, mobile, and desktop trading platforms.
```

Hopefully IBKR can stabilize their platform soon.

**Comment by JackD111** [comment]
Got a response to my ticket regarding the usage of the protal.proxy:

Thank you for contacting Interactive Brokers.

```
It is not recommend to use the Client Portal REST API in a manner not described in the official documentation. As this method is not officially tested or documented we are unable to provide support with any issues you may experience. Please refer to the official documentation for Client Portal API here: LINK../index.html

If you are experiencing issues getting started with the API, or have any other questions, please feel free to get in touch with us again at: LINK../clientservices

Kind Regards,
XX
Clinet Services - Technical Assistance Center, API
```

Sounds to me like the they are fine with it but won't help incase of issues since it's not officially supported. Looks like I'll continue developing my new library :>

**Comment by stopdesign** [comment]
@JackD111
> Sounds to me like the they are fine with it
Good news!

I've tested the direct `portal.proxy` approach and it is much more stable than any gateway. It looks like a single auth session works for at least many hours until IBKR resets everything at the and of the day.

I implemented (partially) their homemade cryptography (https://ndcdyn.interactivebrokers.com/sso/CombineFiles) and authentication logic in Python, so now I'm able to obtain `XYZAB`, `USERID` and `cp` cookies **without** using Chrome, Selenium, Docker and anything like Wallace and Gromit style breakfast machine.

**Comment by JackD111** [comment]
@stopdesign Nice work, mind sharing your auth code? Would save me some time :>

**Comment by 10ne1** [comment]
@stopdesign & @JackD111 That's a very interesting idea but...

The problem with re-implementing parts of IBKR custom logic is they can change it at any time and risks breaking our use case, so that is why I prefer to keep my app using the rust fantoccini crate for webdriver with their original code.

This is the kind of breakage or "unexpected consequence" they warn about, for which they cannot provide any support or guarantee, so I'd rather play it safe and keep auth via a Chrome or Firefox. Honestly the overhead is quite minimal since it's all automated in the background.

**Comment by gbtota** [comment]
@stopdesign Can you  your auth code?

**Comment by gengjun** [comment]
hi @JackD111 and @10ne1 , two years past, have you guys using the portal.proxy directly without the gateway ? how did it go ?

**Comment by tralafiti** [comment]
As a Individual you can use the official web api without any need for the gateway. It's a customized oauth1-process but doable any why robuster than fiddling around with a gateway or hijacking the client portal. 

An introduction can be found under https://www.interactivebrokers.com/campus/ibkr-api-page/webapi-doc/#web-api-access-for-individuals-4 and https://www.interactivebrokers.com/campus/ibkr-api-page/cpapi-v1/#first-party-oauth

Needed credentials can be configured under https://www.interactivebrokers.co.uk/oauth/#/configuration
Technical Details regarding the modified oauth1 https://www.interactivebrokers.com/webtradingapi/oauth.pdf

**Comment by bfoz** [comment]
>  https://www.interactivebrokers.com/campus/ibkr-api-page/cpapi-v1/#first-party-oauth

Did IBKR give you any trouble when you emailed apiintegration@interactivebrokers.com? Did you have to prove that you're not an individual investor?
